x-pack/plugins/xpack_main/server/lib/populate_ui_capabilities.ts

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+import { UICapabilities } from 'ui/capabilities';
+import { Feature } from '../../types';
+
+export function populateUICapabilities(
+  xpackMainPlugin: Record<string, any>,
+  injectedVars: Record<string, any>
+): UICapabilities {
+  const features: Feature[] = xpackMainPlugin.getFeatures();
+
+  const featureCapabilities: UICapabilities[] = features.map(getCapabilitiesFromFeature);
+
+  return mergeCapabilities(injectedVars.uiCapabilities || {}, ...featureCapabilities);
```

Comment:
If we let OSS handle the navLinks for us, it seems like we could simplify the merge logic if we're dealing with a single top-level UI Capability definition per feature. It'd also be nice in the merge logic if we could either throw an explicit error if we're going to overwrite an existing uiCapability definition, or always ignore it if it's already defined.

Reply:
> If we let OSS handle the navLinks for us, it seems like we could simplify the merge logic if we're dealing with a single top-level UI Capability definition per feature.

We _might_ be able to, but I think we'll end up with something similar due to feature catalogue items and management links. I don't have an implementation for them yet, but one or both of those may have dedicated buckets within `uiCapabilities` that need to be merged. Given I don't have an implementation yet, I don't mind simplifying this now, and seeing if I need to complicate it down the road.

> It'd also be nice in the merge logic if we could either throw an explicit error if we're going to overwrite an existing uiCapability definition, or always ignore it if it's already defined.

Sounds good. In an earlier iteration, I was explicitly filtering out definitions that were already defined, but I removed it since the merge order is already taking care of it. Would you prefer to see them ignored, or have an error thrown?